### PR TITLE
HostAuthorization: confusing error message when using multiple hostnames

### PR DESCRIPTION
**Pull request by Eusebius1920**
See #40230

### Feedback needed

This pull request tries to solve the mentioned issue by displaying all hosts that have been blocked by `HostAuthorization`.

**But** there are currently two test cases that are failing, that were fine before the commits of this pull request.

Failing Tests:

* `HostAuthorizationTest#test_detects_localhost_domain_spoofing`
* `HostAuthorizationTest#test_blocks_requests_with_spoofed_X-FORWARDED-HOST`

both located in `actionpack/test/dispatch/host_authorization_test.rb`.

From my understanding these test did **not fail** for the **wrong reasons**. It is in both cases expected that the **permitted, potentially spoofed hosts** are printed in the error message as `Blocked host:`. But they only were printed there, because the other host name within the test were **not permitted** and the bug described in #40230 lead to **always printing the X-FORWARDED-HOST** if any host name was not permitted.

Seeing the **security implications** of this is beyond my understanding, thus I would like to have someone have a look at it, before "fixing" the tests (by me or someone else).

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] There are no typos in commit messages and comments.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Feature branch is up-to-date with `main` (if not - rebase it).
* [x] Pull request only contains one commit for bug fixes and small features. If it's a larger feature, multiple commits are permitted but must be descriptive.
* [x] Tests are added if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.
* [ ] PR is not in a draft state.
* [ ] CI is passing.

### CODE REVIEW

File: actionpack/lib/action_dispatch/middleware/host_authorization.rb [hunk 1]
```diff
@@ -137,16 +142,24 @@ def call(env)
...
-      def authorized?(request)
+      def blocked_hosts(request)
         origin_host = request.get_header("HTTP_HOST")
         forwarded_host = request.x_forwarded_host&.split(/,\s?/)&.last
 
-        @permissions.allows?(origin_host) && (forwarded_host.blank? || @permissions.allows?(forwarded_host))
+        hosts = []
+        hosts << origin_host unless @permissions.allows?(origin_host)
+        hosts << forwarded_host unless forwarded_host.blank? || @permissions.allows?(forwarded_host)
+        hosts
```
Comment (p8) [inline actionpack/lib/action_dispatch/middleware/host_authorization.rb:158]: I think using `reject` here makes the code simpler:
```suggestion
        [origin_host, forwarded_host].compact_blank.reject {|host|  @permissions.allows?(host) }
```

File: actionpack/lib/action_dispatch/middleware/host_authorization.rb [hunk 2]
```diff
@@ -92,11 +92,16 @@ def call(env)
...
+      def blocked_hosts
+        @blocked_hosts || []
+      end
+
       private
         def response_body(request)
           return "" unless request.get_header("action_dispatch.show_detailed_exceptions")
 
-          template = DebugView.new(host: request.host)
+          template = DebugView.new(hosts: blocked_hosts)
```
Comment (p8) [inline actionpack/lib/action_dispatch/middleware/host_authorization.rb:104]: I think this should call the `blocked_hosts` method?
```suggestion
          template = DebugView.new(hosts: blocked_hosts(request))
```

File: actionpack/lib/action_dispatch/middleware/host_authorization.rb (hunk 1 above)
Comment (Eusebius1920) [reply actionpack/lib/action_dispatch/middleware/host_authorization.rb:158]: awesome, thanks!

**Comment by Eusebius1920** [comment]
@MaxLap yes I could change the PR to just do that. I would perfectly be okay with it. Would be only a minor change that could be hacked together in a few spare minutes.

But really blocking this PR is my request for feedback for the failing tests. (See description of this PR) I just don't have the necessary knowledge to evaluate and decide that on my own.

**Comment by MaxLap** [comment]
If your guess is correct about it being caused by the lack of a host that you didn't include because of your extra logic, then I guess my suggestion of always including both it would solve those failing tests.

I can take a look more in depth once the tests run on the edited version.

**Comment by Eusebius1920** [comment]
Yes, you are right. 

I will change the PR such that every host get printed, rebase with upstream and come back to you once done.

**Comment by Eusebius1920** [comment]
I pushed a second variant (all tests are fixed, still there is the issue with the test failing for the wrong reasons - now it is convered under the suceeding tests again).

I personally liked my first attempt more. Now the user gets better while still confusing error messages. All hosts are called "blocked" hosts now. Even if they are not blocked (all). And thus tests are no longer failing, but i keep repeating, see my request for feedback for failing tests. It always comes back to this.

**Comment by MaxLap** [comment]
Sorry if you felt frustration from me not initialy answering what you considered the most important. This is new code for me too, so I wanted some time to think about it.

I looked into it and thought a bit more about this from a security point of view.

I think there is a valid reason for not all hosts to be shown in the error message in the webpage: Always showing the origin_host can mean displaying internal networking information (such as internal DNS adress being used). Showing only `request.host` means showing the publicly visibly host, which is safer.

However, as we know, it's confusing since that may not be the actual cause of the failure.

In the logs, however, we can can provide as much information as we want. So my thinking here is this:

1) In the webpage, only show `request.host` (as it used to do)
2) In the webpage, change the h1 to "Blocked host: <%= @host %> (or another internal host)"
3) In the webpage, add "Your logs will also contain the internal hosts if any" before the "To allow requests...."
4) And in the log, we are free to do what we want. I would personally replace the log message to "Block host, one of: #{all_hosts_here}"

This way, someone isn't purely blind-sided by the presence of a reverse-proxy, we don't reveal internal information and the messages aren't straight up wrong (just not as precise as they could be). To me, this is not confusing anymore.

The reason I personally wouldn't put the extra code to only display the blocked host is to keep the code as simple as possible. Keeping state (like you did with the `@blocked_host`) introduces complexity & risk around things I personally don't know. For example, if there are multiple threads, could a different thread clear or change the value of @blocked_host before it is used, leading to plain wrong messages?

From looking at the two tests you mention, if we consider this idea that the "HOST" is secret information when there is forwarding involved, it make sense that they not fail.

None of this ensures the PR will go through, but I believe in increases the chances at least.

**Comment by Eusebius1920** [comment]
@MaxLap I was certainly not frustrated, but thankful for your feedback!

I will consider yout point about some hostnames being secret.

Do you have any idea why the buildkite-tests are failing? They do not seem to use the updated code, because they are complaining about the `@host`-variable missing in templates, though it was renamed by this pull request to `@hosts`

**Pull request closed without merging**